### DISCUSSION TITLE
Some examples fail to compile on linux

### DISCUSSION THREAD

**Discussion by ivmidable** [uNetworking/uWebSockets#1074] [Q&A] [answered]
examples/EchoServer.cpp: In function ‘int main()’:
examples/EchoServer.cpp:45:6: error: no matching function for call to ‘uWS::TemplatedApp<true>::ws<main()::PerSocketData>(const char [3], <brace-enclosed initializer list>)’
     }).listen(9001, [](auto *token) {
      ^
In file included from examples/EchoServer.cpp:2:
src/App.h:104:20: note: candidate: ‘template<class UserData> uWS::TemplatedApp<SSL>&& uWS::TemplatedApp<SSL>::ws(std::__cxx11::string, uWS::TemplatedApp<SSL>::WebSocketBehavior&&) [with UserData = UserData; bool SSL = true]’
     TemplatedApp &&ws(std::string pattern, WebSocketBehavior &&behavior) {
                    ^~
src/App.h:104:20: note:   template argument deduction/substitution failed:
examples/EchoServer.cpp:45:6: note:   cannot convert ‘{SHARED_COMPRESSOR, (16 * 1024), 10, ((1 * 1024) * 1024), <lambda closure object>main()::<lambda(auto:11*)>{}, <lambda closure object>main()::<lambda(auto:12*, std::string_view, uWS::OpCode)>{}, <lambda closure object>main()::<lambda(auto:13*)>{}, <lambda closure object>main()::<lambda(auto:14*)>{}, <lambda closure object>main()::<lambda(auto:15*)>{}, <lambda closure object>main()::<lambda(auto:16*, int, std::string_view)>{}}’ (type ‘<brace-enclosed initializer list>’) to type ‘uWS::TemplatedApp<true>::WebSocketBehavior&&’
     }).listen(9001, [](auto *token) {
      ^
examples/EchoServerThreaded.cpp: In lambda function:
examples/EchoServerThreaded.cpp:43:14: error: no matching function for call to ‘uWS::TemplatedApp<false>::ws<main()::PerSocketData>(const char [3], <brace-enclosed initializer list>)’
             }).listen(9001, [](auto *token) {
              ^
In file included from examples/EchoServerThreaded.cpp:1:
src/App.h:104:20: note: candidate: ‘template<class UserData> uWS::TemplatedApp<SSL>&& uWS::TemplatedApp<SSL>::ws(std::__cxx11::string, uWS::TemplatedApp<SSL>::WebSocketBehavior&&) [with UserData = UserData; bool SSL = false]’
     TemplatedApp &&ws(std::string pattern, WebSocketBehavior &&behavior) {
                    ^~
src/App.h:104:20: note:   template argument deduction/substitution failed:
examples/EchoServerThreaded.cpp:43:14: note:   cannot convert ‘{SHARED_COMPRESSOR, (16 * 1024), 10, ((1 * 1024) * 1024), <lambda closure object>main()::<lambda(std::thread*)>::<lambda()>::<lambda(auto:11*)>{}, <lambda closure object>main()::<lambda(std::thread*)>::<lambda()>::<lambda(auto:12*, std::string_view, uWS::OpCode)>{}, <lambda closure object>main()::<lambda(std::thread*)>::<lambda()>::<lambda(auto:13*)>{}, <lambda closure object>main()::<lambda(std::thread*)>::<lambda()>::<lambda(auto:14*)>{}, <lambda closure object>main()::<lambda(std::thread*)>::<lambda()>::<lambda(auto:15*)>{}, <lambda closure object>main()::<lambda(std::thread*)>::<lambda()>::<lambda(auto:16*, int, std::string_view)>{}}’ (type ‘<brace-enclosed initializer list>’) to type ‘uWS::TemplatedApp<false>::WebSocketBehavior&&’
             }).listen(9001, [](auto *token) {
              ^
examples/BroadcastingEchoServer.cpp: In function ‘int main()’:
examples/BroadcastingEchoServer.cpp:46:6: error: no matching function for call to ‘uWS::TemplatedApp<false>::ws<main()::PerSocketData>(const char [3], <brace-enclosed initializer list>)’
     }).listen(9001, [](auto *token) {
      ^
In file included from examples/BroadcastingEchoServer.cpp:1:
src/App.h:104:20: note: candidate: ‘template<class UserData> uWS::TemplatedApp<SSL>&& uWS::TemplatedApp<SSL>::ws(std::__cxx11::string, uWS::TemplatedApp<SSL>::WebSocketBehavior&&) [with UserData = UserData; bool SSL = false]’
     TemplatedApp &&ws(std::string pattern, WebSocketBehavior &&behavior) {
                    ^~
src/App.h:104:20: note:   template argument deduction/substitution failed:
examples/BroadcastingEchoServer.cpp:46:6: note:   cannot convert ‘{DEDICATED_COMPRESSOR_3KB, ((16 * 1024) * 1024), 10, ((1 * 1024) * 1024), <lambda closure object>main()::<lambda(auto:11*)>{}, <lambda closure object>main()::<lambda(auto:12*, std::string_view, uWS::OpCode)>{}, <lambda closure object>main()::<lambda(auto:13*)>{}, <lambda closure object>main()::<lambda(auto:14*)>{}, <lambda closure object>main()::<lambda(auto:15*)>{}, <lambda closure object>main()::<lambda(auto:16*, int, std::string_view)>{}}’ (type ‘<brace-enclosed initializer list>’) to type ‘uWS::TemplatedApp<false>::WebSocketBehavior&&’
     }).listen(9001, [](auto *token) {
      ^

**Comment by ghost** [comment, marked as answer]
You probably have a too old compiler, try with GCC 9 or Clang 10

**Comment by hst-m** [comment]
for Ubuntu 20.04 can just do `sudo apt install -y clang build-essential libssl-dev zlib1g-dev` to make everything, examples, benchmarks, uwebsockets, uwebsockets.js, usockets

**Comment by ivmidable** [comment]
Ah crap sorry about that, I was on GCC 8. It's working now.

**Comment by ghost** [comment]
No problem, I need to update the user manual in this regard since it does require C++20
